File: src/parser/nim/parser/tools/formatter.nim
import std/[strformat, strutils, tables, re]

import ../helpers/[types, charsets]

# Formats (prettifies) .acmap file.
#
# @param  {object} S - State object.
# @return {string} - The prettied file contents.
proc formatter*(S: State): tuple =
    let igc = S.args.igc
    var nodes = S.tables.tree["nodes"]
    let eN = Node()
    var output: seq[string] = @[]
    var passed: seq[Node] = @[]
    let r = re"^[ \t]+"
    var alias: Node

    # Indentation level multipliers.
    const MXP = {
        "COMMENT": 0,
        "COMMAND": 0,
        "FLAG": 1,
        "OPTION": 2,
        "BRACE": 0,
        "NEWLINE": 0,
        "SETTING": 0,
        "VARIABLE": 0
    }.toTable

    var nl_count = 0 # Track consecutive newlines.
    var scopes: seq[int] = @[] # Track command/flag scopes.

    let (ichar, iamount) = S.args.fmt
    proc indent(`type`: string = "COMMENT", count: int = 0): string =
        let level = if count > 0: count else: MXP[`type`]
        result = ichar.repeat(level * iamount)

    # Gets next node that is not a comment. Also takes into account
    #     subsequent newline node.
    #
    # @param  {number} i - The index to start search.
    # @param  {number} l - The length of array.
    # @return {object} - The node object.
    proc nextnode(i, l: int): Node =
        result = eN
        if igc:
            var i = i + 1
            while i < l:
                let N = nodes[i]
                let t = N.kind
                if t != nkComment:
                    result = N
                    break
                elif t == nkComment:
                    inc(i)
                inc(i)
        else: result = if i + 1 < l: nodes[i + 1] else: eN

    # Gets the node previously iterated over.
    #
    # @param  {number} i - The index to start search.
    # @param  {number} l - The length of array.
    # @return {object} - The node object.
    proc lastnode(i, l: int): Node =
        result = eN
        if igc: result = passed[passed.high]
        else: result = if i - 1 < l: nodes[i - 1] else: eN

    # Loop over nodes to build formatted file.

    var i = 0
    let l = nodes.len
    while i < l:
        let N = nodes[i]
        let t = N.kind

        # Ignore starting newlines.
        if output.len == 0 and t == nkNewline:
            inc(i)
            continue
        # Remove comments when flag is provided.
        if igc and t == nkComment:
            inc(i) # + 1 to skip next newline node.
            inc(i) # + 1 to account for continue.
            continue

        case (t):
            of nkComment:
                let scope = if scopes.len > 0: scopes[^1] else: 0
                let pad = if not N.inline: indent(count = scope) else: " "

                output.add(fmt"{pad}{N.comment.value}")

            of nkNewline:
                let nN = nextnode(i, l);

                if nl_count <= 1: output.add("\n")
                inc(nl_count)
                if nN.kind != nkNewline: nl_count = 0

                if scopes.len != 0:
                    let last = output[output.len - 2]
                    let lchar = last[last.len - 1]
                    let isbrace = lchar == C_LBRACKET or lchar == C_LPAREN
                    if isbrace and nN.kind == nkNewline: inc(nl_count)
                    if nN.kind == nkBrace:
                        if lastnode(i, l).kind == nkNewline: discard output.pop()

            of nkSetting:
                let nval = N.name.value
                let aval = N.assignment.value
                let vval = N.value.value

                var r = "@"
                if nval != "":
                    r &= nval
                    if aval != "":
                        r &= fmt" {aval}"
                        if vval != "":
                            r &= fmt" {vval}"

                output.add(r)

            of nkVariable:
                let nval = N.name.value
                let aval = N.assignment.value
                let vval = N.value.value

                var r = "$"
                if nval != "":
                    r &= nval
                    if aval != "":
                        r &= fmt" {aval}"
                        if vval != "":
                            r &= fmt" {vval}"

                output.add(r)

            of nkCommand:
                let vval = N.value.value
                let cval = N.command.value
                let dval = N.delimiter.value
                let aval = N.assignment.value

                var r = ""
                if cval != "":
                    r &= cval
                    if dval != "":
                        r &= dval
                    else:
                        if aval != "":
                            r &= fmt" {aval}"
                            if vval != "":
                                r &= fmt" {vval}"

                let nN = nextnode(i, l)
                if nN.kind == nkFlag: r &= " "
                output.add(r)
                if vval != "" and vval == "[": scopes.add(1) # Track scope.

            of nkFlag:
                if N.virtual:
                    inc(i) # + 1 to account for continue.
                    continue

                let kval = N.keyword.value
                let hval = N.hyphens.value
                let nval = N.name.value
                let bval = N.boolean.value
                let aval = N.assignment.value
                let dval = N.delimiter.value
                let mval = N.multi.value
                let vval = N.value.value
                let ival = N.alias.value
                let singleton = N.singleton
                let pad = indent(count = singleton.int)
                var pipe_del = if singleton: "" else: "|"

                # Skip if an alias and set ref for later.
                if ival != "" and ival == nval:
                    alias = N
                    inc(i) # + 1 to account for continue.
                    continue

                # Note: If nN is a flag reset var.
                if pipe_del != "":
                    let nN = nextnode(i, l)
                    if nN.kind != nkFlag: pipe_del = ""

                var r = pad

                if kval != "":
                    r &= kval
                    if vval != "": r &= fmt" {vval}"
                else:
                    if hval != "":
                        r &= hval
                        if nval != "":
                            r &= nval
                            if ival != "" and alias != nil and ival == alias.alias.value:
                                r &= "::" & ival
                                alias = nil
                            if bval != "":
                                r &= bval
                            elif aval != "":
                                r &= aval
                                if mval != "": r &= mval
                                if dval != "": r &= dval
                                if vval != "": r &= vval

                output.add(r & pipe_del)

                if vval != "" and vval == "(": scopes.add(2) # Track scope.

            of nkOption:
                let bval = N.bullet.value
                let vval = N.value.value
                let pad = indent("OPTION")

                var r = pad
                if bval != "":
                    r &= bval
                    if vval != "":
                        r &= fmt" {vval}"

                output.add(r)

            of nkBrace:
                let bval = N.brace.value
                var pad = indent(count = if bval == "]": 0 else: 1)

                if bval == ")":
                    let l = output.len
                    let last = output[l - 1]
                    let slast = output[l - 2]
                    let ll = last.len
                    let lfchar = last.replace(r, "")[0]
                    let slchar = slast[slast.len - 1]

                    if lfchar == C_HYPHEN and last[ll - 1] == C_LPAREN: pad = ""
                    elif last == "\n" and slchar == C_LPAREN:
                        pad = ""
                        discard output.pop()
                elif bval == "]":
                    let l = output.len
                    let last = output[l - 1]
                    let slast = output[l - 2]
                    if last == "\n" and slast == "\n":
                        discard output.pop()
                    else:
                        let sl = slast.len
                        let slchar = slast[sl - 1]
                        if last == "\n" and slchar == C_LBRACKET: discard output.pop()

                output.add(fmt"{pad}{bval}")
                if scopes.len > 0: discard scopes.pop() # Un-track last scope.

            else: discard

        passed.add(N)
        inc(i)

    i = output.high
    while true:
        if output[i] != "\n": break
        discard output.pop()
        dec(i)

    var data: tuple[
        acdef: string,
        config: string,
        keywords: string,
        filedirs: string,
        contexts: string,
        formatted: string,
        placeholders: StringTableRef,
        tests: string
    ]
    data.formatted = output.join("") & "\n"
    result = data

File: src/parser/tools/formatter.js
"use strict";

const { nk } = require("../helpers/enums.js");

/**
 * Formats (prettifies) .acmap file.
 *
 * @param  {object} S - State object.
 * @return {string} - The prettied file contents.
 */
module.exports = (S) => {
	let { fmt, igc } = S.args;
	let { nodes } = S.tables.tree;
	const eN = {};
	let output = [];
	let passed = [];
	const r = /^[ \t]+/g;
	let alias;

	// Indentation level multipliers.
	let MXP = {
		COMMENT: 0,
		COMMAND: 0,
		FLAG: 1,
		OPTION: 2,
		BRACE: 0,
		NEWLINE: 0,
		SETTING: 0,
		VARIABLE: 0
	};

	let nl_count = 0; // Track consecutive newlines.
	let scopes = []; // Track command/flag scopes.

	const [ichar, iamount] = fmt;
	let indent = (type, count) => ichar.repeat((count || MXP[type]) * iamount);

	/**
	 * Gets next node that is not a comment. Also takes into account
	 *     subsequent newline node.
	 *
	 * @param  {number} i - The index to start search.
	 * @param  {number} l - The length of array.
	 * @return {object} - The node object.
	 */
	let nextnode = (i, l) => {
		let r = eN;
		if (igc) {
			for (i = i + 1; i < l; i++) {
				let N = nodes[i];
				let type = N.node;
				if (type !== nk.Comment) {
					r = N;
					break;
				} else if (type === nk.Comment) i++;
			}
		} else r = nodes[i + 1];

		return r;
	};

	/**
	 * Gets the node previously iterated over.
	 *
	 * @param  {number} i - The index to start search.
	 * @param  {number} l - The length of array.
	 * @return {object} - The node object.
	 */
	let lastnode = (i /*, l*/) => {
		let r = eN;
		if (igc) r = passed[passed.length - 1];
		else r = nodes[i - 1];
		return r;
	};

	// Loop over nodes to build formatted file.

	for (let i = 0, l = nodes.length; i < l; i++) {
		let N = nodes[i];
		let type = N.node;

		// Ignore starting newlines.
		if (!output.length && type === nk.Newline) continue;
		// Remove comments when flag is provided.
		if (igc && type === nk.Comment) {
			i++;
			continue;
		}

		switch (type) {
			case nk.Comment:
				{
					let scope = scopes[scopes.length - 1] || null;
					let pad = !N.inline ? indent(null, scope) : " ";

					output.push(`${pad}${N.comment.value}`);
				}

				break;

			case nk.Newline:
				{
					let nN = nextnode(i, l);

					if (nl_count <= 1) output.push("\n");
					nl_count++;
					if (nN && nN.node !== nk.Newline) nl_count = 0;

					if (scopes.length) {
						let last = output[output.length - 2];
						let lchar = last[last.length - 1];
						let isbrace = lchar === "[" || lchar === "(";
						if (isbrace && nN && nN.node === nk.Newline) nl_count++;
						if (nN.node === nk.Brace) {
							if (lastnode(i, l).node === nk.Newline)
								output.pop();
						}
					}
				}

				break;

			case nk.Setting:
				{
					let nval = N.name.value;
					let aval = N.assignment.value;
					let vval = N.value.value;

					let r = "@";
					if (nval) {
						r += nval;
						if (aval) {
							r += ` ${aval}`;
							if (vval) {
								r += ` ${vval}`;
							}
						}
					}

					output.push(r);
				}

				break;

			case nk.Variable:
				{
					let nval = N.name.value;
					let aval = N.assignment.value;
					let vval = N.value.value;

					let r = "$";
					if (nval) {
						r += nval;
						if (aval) {
							r += ` ${aval}`;
							if (vval) {
								r += ` ${vval}`;
							}
						}
					}

					output.push(r);
				}

				break;

			case nk.Command:
				{
					let vval = N.value.value;
					let cval = N.command.value;
					let dval = N.delimiter.value;
					let aval = N.assignment.value;

					let r = "";
					if (cval) {
						r += cval;
						if (dval) {
							r += dval;
						} else {
							if (aval) {
								r += ` ${aval}`;
								if (vval) {
									r += ` ${vval}`;
								}
							}
						}
					}

					let nN = nextnode(i, l);
					if (nN && nN.node === nk.Flag) r += " ";
					output.push(r);
					if (vval && vval === "[") scopes.push(1); // Track scope.
				}

				break;

			case nk.Flag:
				{
					if (N.virtual) {
						continue;
					}

					let kval = N.keyword.value;
					let hval = N.hyphens.value;
					let nval = N.name.value;
					let bval = N.boolean.value;
					let aval = N.assignment.value;
					let dval = N.delimiter.value;
					let mval = N.multi.value;
					let vval = N.value.value;
					let ival = N.alias.value;
					let singleton = N.singleton;
					let pad = indent(null, singleton ? 1 : null);
					let pipe_del = singleton ? "" : "|";

					// Skip if an alias and set ref for later.
					if (ival && ival === nval) {
						alias = N;
						continue;
					}

					// Note: If nN is a flag reset var.
					if (pipe_del) {
						let nN = nextnode(i, l);
						if (nN && nN.node !== nk.Flag) pipe_del = "";
					}

					// [https://stackoverflow.com/a/23867090]
					let r = pad;

					if (kval) {
						r += kval;
						if (vval) r += ` ${vval}`;
					} else {
						if (hval) {
							r += hval;
							if (nval) {
								r += nval;
								if (
									ival &&
									alias &&
									ival === alias.alias.value
								) {
									r += `::${ival}`;
									alias = null;
								}
								if (bval) {
									r += bval;
								} else if (aval) {
									r += aval;
									if (mval) r += mval;
									if (dval) r += dval;
									if (vval) r += vval;
								}
							}
						}
					}

					output.push(r + pipe_del);

					if (vval && vval === "(") scopes.push(2); // Track scope.
				}

				break;

			case nk.Option:
				{
					let bval = N.bullet.value;
					let vval = N.value.value;
					let pad = indent("OPTION");

					let r = pad;
					if (bval) {
						r += bval;
						if (vval) {
							r += ` ${vval}`;
						}
					}

					output.push(r);
				}

				break;

			case nk.Brace:
				{
					let bval = N.brace.value;
					let pad = indent(null, bval === "]" ? 0 : 1);

					if (bval === ")") {
						let l = output.length;
						let last = output[l - 1];
						let slast = output[l - 2];
						let ll = last.length;
						let lfchar = last.replace(r, "")[0];
						let slchar = slast[slast.length - 1];

						if (lfchar === "-" && last[ll - 1] === "(") pad = "";
						else if (last === "\n" && slchar === "(") {
							pad = "";
							output.pop();
						}
					} else if (bval === "]") {
						let l = output.length;
						let last = output[l - 1];
						let slast = output[l - 2];
						if (last === "\n" && slast === "\n") {
							output.pop();
						} else {
							let sl = slast.length;
							let slchar = slast[sl - 1];
							if (last === "\n" && slchar === "[") output.pop();
						}
					}

					output.push(`${pad}${bval}`);
					scopes.pop(); // Un-track last scope.
				}

				break;
		}

		passed.push(N);
	}

	/*eslint no-constant-condition: ["error", { "checkLoops": false }]*/
	let i = output.length - 1;
	while (true) {
		if (output[i] !== "\n") break;
		output.pop();
		i--;
	}

	return output.join("") + "\n";
};
